### PR TITLE
Document //flutter/shell/common/engine.

### PR DESCRIPTION
**Pull request by chinmaygarde**

### CODE REVIEW

File: shell/common/engine.h [hunk 1]
```diff
@@ -29,36 +29,241 @@
...
+/// The root isolate of Flutter application gets "window" bindings. Using these
+/// bindings, the application can schedule frames, post layer-trees for
+/// rendering, ask to decompress images and upload them to the GPU, etc..
+/// Non-root isolates of the VM do not get any of these capabilities and are run
+/// in a VM managed thread pool (so if they did have "window", the threading
+/// guarantees needed for engine operation would be violated).
+///
+/// The engine is responsible for the entire life-cycle of the root isolate.
+/// When the engine is collected, its owner assumes that the root isolate has
+/// been shutdown and appropriate resource collected. While each engine instance
```
Comment (cbracken) [inline shell/common/engine.h]: nit: /resource/resources/

File: shell/common/engine.h [hunk 2]
```diff
@@ -29,36 +29,241 @@
...
+/// can only manage a single instance of a root isolate, it may restart that
+/// isolate on request. This is how the cold-restart development scenario is
+/// supported.
+///
+/// When the engine instance is initially created, the root isolate is created
+/// but it is not in the |DartIsolate::Phase::Running| phase yet. It only moves
+/// into that phase when a successful call to Engine::Run is made.
```
Comment (cbracken) [inline shell/common/engine.h]: |Engine::Run|

File: shell/common/engine.h [hunk 3]
```diff
@@ -29,36 +29,241 @@
...
+///
+/// @note     This name of this class is perhaps a bit unfortunate and has
+///           sometimes been the cause of confusion. For a class named "Engine"
+///           in the Flutter "Engine" repository, its responsibilities are
+///           decidedly unremarkable. But, it does happen to be the primary
+///           entry-point used by components higher up in the Flutter tech stack
+///           (usually in Dart code) to peer into the lower level functionality.
+///           Besides, the authors haven't been able to come up with a more apt
+///           name and it does happen to be one of the older classes in the
+///           repository.
```
Comment (cbracken) [inline shell/common/engine.h]: :)

File: shell/common/engine.h [hunk 4]
```diff
@@ -29,36 +29,241 @@
...
+  ///
   class Delegate {
    public:
+    //--------------------------------------------------------------------------
+    /// @brief      When the accessibility tree has been updated by the Flutter
+    ///             application, this new information needs to be conveyed to
+    ///             the underlying platform. The engine delegates this task to
+    ///             the shell via this call. The engine cannot access the
+    ///             underlying platform directly because of threading
+    ///             considerations. Most platform specific API's to convey
```
Comment (cbracken) [inline shell/common/engine.h]: APIs

File: shell/common/engine.h [hunk 5]
```diff
@@ -69,59 +274,429 @@ class Engine final : public RuntimeDelegate {
...
+  ///             number of these profiles have been gathered, they need to be
+  ///             reported back to Dart code. The shell reports this extra
+  ///             instrumentation information back to Dart code running on the
+  ///             engine by invoking this method at predefined intervals.
+  ///
+  /// @see        `FrameTiming`
+  ///
+  //  TODO(chinmaygarde): The use `int64_t` is added for ease of conversion to
+  //  Dart but hurts readability. The phases and the units of the timepoints are
+  //  not obvious without some sleuthing. The conversion can happen at th native
```
Comment (cbracken) [inline shell/common/engine.h]: th -> the

File: shell/common/engine.h [hunk 6]
```diff
@@ -69,59 +274,429 @@ class Engine final : public RuntimeDelegate {
...
+  ///             the native process (and perhaps return an exit code) because
+  ///             it does not have that power. After all, Flutter applications
+  ///             reside within a host process that may have other
+  ///             responsibilities besides just running Flutter applications.
+  ///             Also, the `main` entry-points are run on an event loop and
+  ///             returning from "main" (unlike in C/C++ applications) does not
+  ///             mean termination of the process. Besides, the return value of
+  ///             the main entrypoint is discarded.
+  ///
+  ///             Once technique used by embedders to determine "liveness" is to
```
Comment (cbracken) [inline shell/common/engine.h]: s/Once/One/

File: shell/common/engine.h [hunk 7]
```diff
@@ -69,59 +274,429 @@ class Engine final : public RuntimeDelegate {
...
+  ///             set a "return value" and that value if present. When the first
+  ///             item of the pair is false, second item is meaningless.
+  ///
   std::pair<bool, uint32_t> GetUIIsolateReturnCode();
 
+  //----------------------------------------------------------------------------
+  /// @brief      Indicates to the Flutter application that it has obtained a
+  ///             rendering surface. This is a good opportunity for the engine
+  ///             to start servicing any outstanding frame requests from the
+  ///             Flutter application. Flutter application that have no
```
Comment (cbracken) [inline shell/common/engine.h]: applications

File: shell/common/engine.h [hunk 8]
```diff
@@ -69,59 +274,429 @@ class Engine final : public RuntimeDelegate {
...
+  //----------------------------------------------------------------------------
+  /// @brief      Notifies the engine that the embedder encountered an
+  ///             accessibility related action on the specified node. This call
+  ///             originates on the platform view and has been forwarded to the
+  ///             engine here on the UI task runner by the shell.
+  ///
+  /// @param[in]  id      The identifier of the accessibility node.
+  /// @param[in]  action  The accessibility related action performed on the
+  ///                     node of the specified ID.
+  /// @param[in]  args    An optional list of argument that apply to the
```
Comment (cbracken) [inline shell/common/engine.h]: arguments

Though really it's just whatever arbitrary blob of serialised data we decided to associate with the action, so maybe just "Optional data that applies to the specified action."

File: shell/common/engine.h (hunk 1 above)
Comment (chinmaygarde) [reply shell/common/engine.h]: Done.

File: shell/common/engine.h (hunk 2 above)
Comment (chinmaygarde) [reply shell/common/engine.h]: Done.

File: shell/common/engine.h (hunk 4 above)
Comment (chinmaygarde) [reply shell/common/engine.h]: Done.

File: shell/common/engine.h (hunk 5 above)
Comment (chinmaygarde) [reply shell/common/engine.h]: Done.

File: shell/common/engine.h (hunk 6 above)
Comment (chinmaygarde) [reply shell/common/engine.h]: Done.

File: shell/common/engine.h (hunk 7 above)
Comment (chinmaygarde) [reply shell/common/engine.h]: Done.

File: shell/common/engine.h (hunk 8 above)
Comment (chinmaygarde) [reply shell/common/engine.h]: Done.

**Pull request merged**